core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java

```diff
@@ -20,9 +20,19 @@
 package org.elasticsearch.common.inject;
 
 /**
+ * This interface can be added to a Module to spawn sub modules. DO NOT USE.
  *
+ * This is fundamentally broken.
+ * <ul>
+ * <li>If you have a plugin with multiple modules, return all the modules at once.</li>
+ * <li>If you are trying to make the implementation of a module "pluggable", don't do it.
+ * This is not extendable because custom implementations (using onModule) cannot be
+ * registered before spawnModules() is called.</li>
+ * </ul>
+ * @deprecated Create modules directly
  */
-public interface SpawnModules {
+@Deprecated
```

Comment:
The logic is not broken. This interface is broken, for the reasons described in the javadoc. All this interface does is make for (1) more classes (2) more obfuscation and (3) possibility for broken pluggability APIs.  If this interface stays around, nothing stops new extension points from being added which are broken again by trying to plugin a module "implementation", which as explained, simply cannot work with plugin isolation.

I don't know what you mean by "module pre processing", but extension points should be clear, and plugins should bind ancillary classes they need in their own modules. For example, if a plugin wants to bind a custom transport protocol, it should add the transport protocol to TransportModule with onModule, and any ancillary classes it needs bound can be added in its own module. There is no need to "group" these bindings together in one module.